### PR TITLE
Expression Tree Balancing

### PR DESCRIPTION
**Pull request by pranavb-ca**
This patch adds a pass to balance expression trees.

We use this pass just before code generation for HVX. I am putting this
in a file of its own so that it can be used for other targets that
may find this useful.

@dsharletg, @ronlieb, @dpalermo : PTAL.

### CODE REVIEW

File: src/BalanceExpressionTrees.cpp [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+    // ExprHeights(Scope<int> var_heights) : var_heights(var_heights) {}
+    void clear() { m.clear(); }
+    void push(const std::string &name, int ht) {
+        var_heights.push(name, ht);
+    }
+    void pop(const std::string &name) {
+        var_heights.pop(name);
+    }
+    void push(Expr e) {
+        internal_assert(e.type().is_vector()) << "We are interested in the heights of only vector types\n";
```
Comment (pranavb-ca) [inline src/BalanceExpressionTrees.cpp]: You are right. This is a vestige from the time that I had written this specifically for HVX and at that time I was trying to limit the scope of this.

File: src/BalanceExpressionTrees.cpp [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+        }
+    }
+
+    void visit(const Call *op) {
+        if (op->type.is_vector()) {
+            // ht(slice_vector(concat_vectors(x, ..)) = ht(concat_vectors(x, ...))
+            if (op->is_intrinsic(Call::slice_vector)) {
+                const Call *concat_v = op->args[0].as<Call>();
+                if (concat_v && concat_v->is_intrinsic(Call::concat_vectors)) {
+                    int ht = height(op->args[0]);
```
Comment (pranavb-ca) [inline src/BalanceExpressionTrees.cpp]: It is taking the height of the concat_vector in itself. "op" is slice_vector.

**Comment by pranavb-ca** [comment]
@dsharletg: I am working on your suggestions, but I am out on vacation starting tonight. So, I suggest we keep this aside and I'll pick up on this once I am back in Dec.

File: src/BalanceExpressionTrees.cpp [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+            m[op] = *std::max_element(heights.begin(), heights.end());
+        }
+    }
+
+};
+
+class FindRoots : public IRVisitor {
+    using IRVisitor::visit;
+
+    bool is_associative_or_cummutative(Expr a) {
```
Comment (steven-johnson) [inline src/BalanceExpressionTrees.cpp]: Spelling: "commutative"

**Comment by pranavb-ca** [comment]
@dsharletg : I have pushed up changes. Please review when you have a chance.

**Comment by steven-johnson** [comment]
I think this could really use some more extensive test cases; AFAICT only one trivial test has been added.

**Comment by pranavb-ca** [comment]
@steven-johnson : fair point. I'll work on adding some tests.

**Comment by SanderVocke** [comment]
Hello Pranav,

I know it is probably not in the scope of this feature, but commenting on the following text in your pass's .h file:

`/* We can generate better code if imbalanced tress of associative and commutative 
   operations are balanced.
   For example, the expression tree for a + 4*b + 6*c + 4*d + e is

                        +
                      /   \
                    /      e
                  +
                /   \
              /       \
            +          *
          /   \       /  \
        /       \    d    4
      +          *
    /   \      /   \
  a      *    c     6
       /   \
      b     4

  This pass converts this tree to
                 +
               /   \
             /       \  
           /           \
         +               +
       /   \           /   \
     *      *         *      +
   /   \   / \      /   \   /  \
  b     4 c   6    d     4 a    e

  On targets such as HVX, the balanced tree above leads to very good code generation by the use
  of multiply-accumulate instructions.
*/`

**Comment by steven-johnson** [comment]
This PR is pretty old (but seemed interesting at the time). Is it ever going to be revisited?

**Comment by pranavb-ca** [comment]
@steven-johnson  - one of my coworkers is interested in this pass and might dedicate sometime to it. However, in the meantime, I'll close this PR and if his experiments, he'll reincarnate this in a different form.

**Pull request closed without merging**